stdlib/Statistics/src/Statistics.jl

```diff
@@ -963,11 +1186,87 @@ julia> quantile(skipmissing([1, 10, missing]), 0.5)
 5.5
 ```
 """
-quantile(itr, p; sorted::Bool=false) = quantile!(collect(itr), p, sorted=sorted)
+quantile(itr, p; sorted::Bool=false, weights::Union{AbstractArray,Nothing}=nothing) =
+    _quantile(itr, p, sorted, weights)
+
+_quantile(itr, p, sorted::Bool, weights::Nothing) =
+    quantile!(collect(itr), p, sorted=sorted)
+
+_quantile(itr::AbstractArray, p, sorted::Bool, weights::Nothing) =
+    quantile!(sorted ? itr : Base.copymutable(itr), p; sorted=sorted)
+
+function _quantile(v::AbstractArray{V}, p, sorted::Bool, w::AbstractArray{W}) where {V,W}
+    # checks
+    isempty(v) && throw(ArgumentError("quantile of an empty array is undefined"))
+    isempty(p) && throw(ArgumentError("empty quantile array"))
+    all(x -> 0 <= x <= 1, p) || throw(ArgumentError("input probability out of [0,1] range"))
+
+    wsum = sum(w)
+    wsum == 0 && throw(ArgumentError("weight vector cannot sum to zero"))
+    length(v) == length(w) || throw(ArgumentError("data and weight vectors must be the same size," *
```

Comment:
If you need to split it over multiple lines then you might want to just use an `if`?